core-rust/mesh-api-server/src/mesh_api/handlers/block.rs

```diff
@@ -0,0 +1,190 @@
+    let block = models::Block {
+        block_identifier: Box::new(to_mesh_api_block_identifier_from_state_version(
+            state_version,
+        )?),
+        parent_block_identifier: Box::new(to_mesh_api_block_identifier_from_state_version(
+            previous_state_version,
+        )?),
+        timestamp: transaction_identifiers.proposer_timestamp_ms,
+        transactions: vec![transaction],
+        metadata: None,
+    };
+
+    // see https://docs.cdp.coinbase.com/mesh/docs/models#blockresponse
+    Ok(Json(models::BlockResponse {
+        block: Some(Box::new(block)),
+        other_transactions: None,
+    }))
+}
+
+pub(crate) async fn handle_block_transaction(
+    state: State<MeshApiState>,
+    Json(request): Json<models::BlockTransactionRequest>,
+) -> Result<Json<models::BlockTransactionResponse>, ResponseError> {
+    assert_matching_network(&request.network_identifier, &state.network)?;
+
+    let database = state.state_manager.database.snapshot();
+    let mapping_context = MappingContext::new(&state.network);
+
+    let state_version =
+        extract_state_version_from_mesh_api_block_identifier(&request.block_identifier)
+            .map_err(|err| err.into_response_error("block_identifier"))?;
+
+    let transaction_identifiers = database
+        .get_committed_transaction_identifiers(state_version)
+        .ok_or_else(|| {
+            ResponseError::from(ApiError::TransactionNotFound).with_details(format!(
+                "Failed fetching transaction identifiers for state version {}",
+                state_version.number()
+            ))
+        })?;
+
+    let (operations, transaction_identifier) = get_operations_and_transaction_identifier(
+        &mapping_context,
+        database.deref(),
+        state_version,
+        &transaction_identifiers,
+        Some(&request.transaction_identifier.hash),
+    )?;
+
+    // see https://docs.cdp.coinbase.com/mesh/docs/models#transaction
+    let transaction = models::Transaction {
+        transaction_identifier: Box::new(models::TransactionIdentifier::new(
+            transaction_identifier,
+        )),
+        operations,
+        related_transactions: None,
+        metadata: None,
+    };
+
+    // see https://docs.cdp.coinbase.com/mesh/docs/models#blocktransactionresponse
+    Ok(Json(models::BlockTransactionResponse {
+        transaction: Box::new(transaction),
+    }))
+}
+
+fn get_operations_and_transaction_identifier(
+    mapping_context: &MappingContext,
+    database: &StateManagerDatabase<impl ReadableRocks>,
+    state_version: StateVersion,
+    transaction_identifiers: &CommittedTransactionIdentifiers,
+    requested_transaction_identifier: Option<&str>,
+) -> Result<(Vec<models::Operation>, String), ResponseError> {
+    let (operations, transaction_identifier) = match transaction_identifiers
+        .transaction_hashes
+        .as_user()
+    {
+        // In case of non-user transaction we return empty transaction,
```

Comment:
Can you expand this comment - why does the `mesh-cli` return an error?
* Does it not expect to see transactions which weren't submitted?
* Does it not understand what we return? (i.e. it's an implementation bug)
* Does some other invariant get broken?

Reply:
According to the description in https://docs.cdp.coinbase.com/mesh/docs/models#blockresponse blocks must form a canonical, CONNECTED chain of blocks.
So we have two options here:

1. In case of non-user transaction we return empty transaction, to preserve block continuity.

2. [Presumably] We could return None for non-user transactions, if `parent_block_identifier` for user transaction was pointing to the previous user transaction.
This one seems really nice, but we would need some mapping from block_identifier to state version. 

I would suggest going with 1) for now.

Reply:
@lrubasze  the linked documentation is about blocks not being commited. In Radix, all blocks/state versions are committed...

Are either of these two approaches viable?

1. We return system transactions -- some of these will still have operations, and they would need to be returned to ensure that resource balance changes can be reconciled.
2. If we have to filter out system transactions, we return blocks for state versions with system transactions as blocks with 0 transactions.